src/schema/__tests__/FroidSchema.test.ts

```diff
@@ -0,0 +1,1690 @@
+
+        type Query {
+          node(id: ID!): Node
+        }
+
+        interface Node {
+          id: ID!
+        }
+
+        type User implements Node @key(fields: "userId customField1 customField2") {
+          id: ID!
+          userId: String!
+          customField1: UsedCustomScalar1
+          customField2: [UsedCustomScalar2!]!
+        }
+
+        type Todo implements Node @key(fields: "todoId customField") {
+          id: ID!
+          todoId: Int!
+          customField: UsedCustomScalar1
+        }
+      `
+    );
+  });
+
+  describe('when using contacts with @tag', () => {
+    it('propogates valid tags to all core relay object identification types', () => {
+      const productSchema = gql`
+        type Query {
+          topProducts(first: Int = 5): [Product]
+        }
+
+        type Product @key(fields: "upc") {
+          upc: String!
+          name: String
+          price: Int
+          weight: Int
+        }
+      `;
+      const subgraphs = new Map();
+      subgraphs.set('product-subgraph', productSchema);
+
+      const actual = generateSchema({
+        subgraphs,
+        froidSubgraphName: 'relay-subgraph',
+        contractTags: ['storefront', 'supplier'],
+      });
+
+      expect(actual).toEqual(
+        // prettier-ignore
+        gql`
+          extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+          type Query {
+            node(id: ID!): Node @tag(name: "storefront") @tag(name: "supplier")
+          }
+
+          interface Node @tag(name: "storefront") @tag(name: "supplier") {
+            id: ID!
+          }
+
+          type Product implements Node @key(fields: "upc") {
+            id: ID!
+            upc: String!
+          }
+        `
+      );
+    });
+
+    it('uses the first non-id key directive despite contract tags', () => {
+      const productSchema = gql`
+        type Query {
+          topProducts(first: Int = 5): [Product]
+        }
+
+        type Product @key(fields: "upc") @key(fields: "name") {
+          upc: String! @inaccessible
+          name: String @tag(name: "storefront")
+          price: Int
+          weight: Int
+        }
+      `;
+      const subgraphs = new Map();
+      subgraphs.set('product-subgraph', productSchema);
+
+      const actual = generateSchema({
+        subgraphs,
+        froidSubgraphName: 'relay-subgraph',
+        contractTags: ['storefront', 'supplier'],
+      });
+
+      expect(actual).toEqual(
+        // prettier-ignore
+        gql`
+          extend schema @link(url: "https://specs.apollo.dev/federation/v2.0", import: ["@key", "@tag", "@external"])
+
+          type Query {
+            node(id: ID!): Node @tag(name: "storefront") @tag(name: "supplier")
+          }
+
+          interface Node @tag(name: "storefront") @tag(name: "supplier") {
+            id: ID!
+          }
+
+          type Product implements Node @key(fields: "upc") {
```

Comment:
I don't get this one

Reply:
This was a preexisting test I brought over to ensure parity (outside where I expected changes in results to occur). I _think_ this is testing that the `id` field gets the correct tag regardless of overall tags. Meaning, `Product.id` should _only_ be `storefront` whereas the `Node` interface is both `storefront` and `supplier`. I'll update the test description.

Reply:
This is testing that the key used ignores tagging and the resulting `id` honors the overall tagging of the entity itself. So `Product.upc` is used as the key even though it is `@inaccessible` while `Product.id` is tagged `storefront` matching the `Product.name`.